### PR TITLE
Default super().__torch_dispatch__ to one hop dispatch if composite available

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #83012
* #83010
* __->__ #82976

Today, if you have a CompositeImplicitAutograd kernel but it didn't get
applied (because, e.g., inference mode was enabled), once you get to the
`__torch_dispatch__` handler you have no way of saying "well actually, I
wanted the composite".  This patch makes super().__torch_dispatch__, by
default, attempt a decomposition (if it exists), before permanently
disabling the Python keys and running the conventional kernel.

The way it works is copying the same strategy as
https://github.com/pytorch/pytorch/pull/74728 for "skip one hop".
Instead of adding Python to the exclude set (which will prevent us from
dispatching back to this key), we set a boolean TLS that says "don't
process the next Python key.)  When we see this skip bool is set,
we set it to false, and redispatch past the Python key.  This will take
you to the newly added PythonComposite key, which will either (1) run
the CompositeImplicitAutograd decomposition, with the redispatches
calling into Python, because the skip bool is no longer set, or (2)
exclude Python/PythonComposite permanently and drop into the true
backend implementation.

Not in this PR yet, but PythonComposite offers a last chance for us to
check for wrapper subclasses before going into backend code.  Wrapper
subclasses cannot be validly operated on by, e.g., CPU/CUDA kernels, so
erroring here could help prevent segfaults.  There is one part missing
from this PR, which is if you have a tensor subclass that isn't supported
in the backend, how can you tell if it is safe to attempt the composite
or not?  Right now you can just ask if there is a
CompositeImplicitAutograd kernel registered for this operator, but this
is a bit low level.

We discussed a number of alternate strategies, which also seem viable.
This one just seemed the simplest to implement.

- We could make it possible to directly call a kernel at a specific
  dispatch key from Python.  In this case, you could test if
  CompositeImplicitAutograd kernel exists, and if so, call it directly.
  I didn't like this because (1) when you call the kernel directly, you
  are obligated to say what the dispatchKeySet it was called at is, but
  this is undefined if you're directly calling a kernel, and (2) it
  seems like a breakage of encapsulation; normally you have to go
  through the dispatcher to work out what kernel to call next, and this
  lets you bypass it entirely.

- We could directly make the C++ composite functions available from
  Python, and then call them directly when we realize they're available.
  This is annoying to implement since operator registration is not fixed
  (it can happen on dynamic library load) so you need a full listener
  infrastructure to do this properly.

TODO: make modes also work this way

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

### CODE REVIEW

File: aten/src/ATen/core/PythonFallbackKernel.cpp [hunk 1]
```diff
@@ -43,10 +43,23 @@ struct StashTLSOnEntryGuard {
...
+}
+
+void pythonFallback(const c10::OperatorHandle& op, c10::DispatchKeySet ks, torch::jit::Stack* stack) {
   TORCH_INTERNAL_ASSERT(tls_on_entry.has_value());
   // c10::impl::ForceDispatchKeyGuard dispatcher_guard(tls_on_entry.value());
   // StashTLSOnEntryGuard stash_guard;
+
+  // TODO: use guard semantics instead, see https://github.com/pytorch/pytorch/pull/74728
+  if (at::impl::TorchDispatchModeTLS::exchange_skip_next(false)) {
+    op.redispatchBoxed(ks.remove(c10::DispatchKey::Python), stack);
```
Comment (bdhirsh) [inline aten/src/ATen/core/PythonFallbackKernel.cpp:59]: Normally when you call `redispatch(ks, ...)`, the intent is to redispatch to "whatever the next-highest-priority kernel is".

In this case though, we basically know that we want to hit the `PythonComposite` kernel. Do you think it's worth making this `redispatchBoxed(DispatchKeySet(DispatchKey::PythonComposite), ...)` just to make intent clearer?

Comment (ezyang) [reply aten/src/ATen/core/PythonFallbackKernel.cpp:59]: Fair enough

File: aten/src/ATen/core/PythonFallbackKernel.cpp [hunk 2]
```diff
@@ -43,10 +43,23 @@ struct StashTLSOnEntryGuard {
...
+      c10::DispatchKeySet({c10::DispatchKey::Python, c10::DispatchKey::PythonComposite}));
+  op.callBoxed(stack);
+}
+
+void pythonFallback(const c10::OperatorHandle& op, c10::DispatchKeySet ks, torch::jit::Stack* stack) {
   TORCH_INTERNAL_ASSERT(tls_on_entry.has_value());
   // c10::impl::ForceDispatchKeyGuard dispatcher_guard(tls_on_entry.value());
   // StashTLSOnEntryGuard stash_guard;
+
+  // TODO: use guard semantics instead, see https://github.com/pytorch/pytorch/pull/74728
```
Comment (albanD) [inline aten/src/ATen/core/PythonFallbackKernel.cpp:57]: How is the current PR working given that this is not restored after this call?

File: aten/src/ATen/core/PythonFallbackKernel.cpp (hunk 1 above)
Comment (albanD) [reply aten/src/ATen/core/PythonFallbackKernel.cpp:59]: Isn't that ok only if ks doesn't contain any key below PythonComposite ? What would happen if the PythonComposite kernel redispatch as well?

File: aten/src/ATen/core/PythonFallbackKernel.cpp (hunk 2 above)
Comment (ezyang) [reply aten/src/ATen/core/PythonFallbackKernel.cpp:57]: exchange modifies it back to false. This makes it match @peterbell10's current PR although I'm presently trying to convince him to switch it to dynamic scoping.

File: aten/src/ATen/core/PythonFallbackKernel.cpp (hunk 1 above)
Comment (ezyang) [reply aten/src/ATen/core/PythonFallbackKernel.cpp:59]: In this case it would definitely be safe because I don't use the passed in dispatch key set in PythonComposite, and you would revert back to consulting the TLS. But I'm also happy with keeping the logic here as is.

File: aten/src/ATen/core/PythonFallbackKernel.cpp [hunk 3]
```diff
@@ -43,10 +43,23 @@ struct StashTLSOnEntryGuard {
   c10::impl::LocalDispatchKeySet saved_;
 };
 
-void pythonFallback(const c10::OperatorHandle& op, torch::jit::Stack* stack) {
+void pythonCompositeFallback(const c10::OperatorHandle& op, c10::DispatchKeySet ks, torch::jit::Stack* stack) {
+  c10::impl::ExcludeDispatchKeyGuard guard_(
+      c10::DispatchKeySet({c10::DispatchKey::Python, c10::DispatchKey::PythonComposite}));
```
Comment (albanD) [inline aten/src/ATen/core/PythonFallbackKernel.cpp:48]: Missing the `PythonTLSSnapshot` key here.

**Comment by albanD** [comment]
Hey!
So we had a long discussion about this yesterday with Brian and we think there are a couple issues with this.

At the highest level, it feels like you now have two ways to access decompositions: via a decomposition table or via `super()`. It is not very clear which one you would want to use and when. And when you want to use both, there is going to be some weird logic.

On the other hand, it feels like having a single mechanism to access decomposition based on a table would be more natural. That table can be extended with labels for decomposition if the user want to easily select/un-select specific decomposition based on their properties.

To address your concern about having to build a listener system for this, wouldn't lazily querying the dispatcher from the decomposition table when the user asks for the decomposition work (at least for now)?

cc @bdhirsh does this capture the conclusion accurately?

File: aten/src/ATen/core/PythonFallbackKernel.cpp (hunk 2 above)
Comment (albanD) [reply aten/src/ATen/core/PythonFallbackKernel.cpp:57]: I am not sure this is sound though as the user has complete control of what happens between when you set that bool and when you unset it. A small modification of your code sample (disable inference mode inside) leads to pretty weird results:
```python
        called = []
        level = 0

        class SubTensor(torch.Tensor):
            __torch_function__ = torch._C._disabled_torch_function_impl

            @classmethod
            def __torch_dispatch__(cls, func, types, args=(), kwargs=None):
                nonlocal level
                print(f"In {func} at {level} with {torch.is_inference_mode_enabled()}")
                called.append((func, level))
                level += 1
                try:
                    # I've seen what I wanted, enable decomps again
                    with torch.inference_mode(False):
                        return super().__torch_dispatch__(func, types, args, kwargs)
                finally:
                    level -= 1

        x = torch.randn(2, 3)
        y = torch.randn(5, 3)
        with torch.inference_mode(True):
            r = torch.nn.functional.linear(SubTensor(x), SubTensor(y))
```
Will generate
```
[(<OpOverload(op='aten.linear', overload='default')>, 0), (<OpOverload(op='aten.mm', overload='default')>, 1)]
```
Basically:
- The second linear is decomposed at the autograd level
- The torch_dispatch to transpose is the one that get skipped by the skip-one boolean
- matmul is decomposed at the autograd level
- mm is captured

**Comment by ezyang** [comment]
this pr is subsumed by https://github.com/pytorch/pytorch/pull/83012
